src/client.ts

```diff
@@ -1,149 +1,224 @@
-// opaque types for each resource, to communicate the records that are expected
-// to come from an endpoint and are provided to iterating functions.
-
-/*
-import { Opaque } from 'type-fest';
-export type AcmeUser = Opaque<any, 'AcmeUser'>;
-export type AcmeGroup = Opaque<any, 'AcmeGroup'>;
-*/
-
-/**
- * An APIClient maintains authentication state and provides an interface to
- * third party data APIs.
- *
- * It is recommended that integrations wrap provider data APIs to provide a
- * place to handle error responses and implement common patterns for iterating
- * resources.
- */
 export class APIClient {
-  constructor(readonly config: IntegrationConfig) {}
+  private readonly clientId: string;
+  private readonly clientSecret: string;
+  private readonly accountId: string;
+  /*
+    Explanation: the reason this accessToken variable is static class variable 
+    is that we don't want each step function's run to request the same (new) token
+    every time. This way, once accessToken value is fetched, it's shared between
+    all the step functions. 
+  */
+  private static accessToken: string;
+
+  constructor(readonly config: CalculatedIntegrationConfig) {
+    this.accountId = config.accountId;
+    this.clientId = config.clientId;
+    this.clientSecret = config.clientSecret;
+  }
 
-  public async verifyAuthentication(): Promise<void> {
-    // TODO make the most light-weight request possible to validate
-    // authentication works with the provided credentials, throw an err if
-    // authentication fails
-    const request = new Promise((resolve, reject) => {
-      http.get(
-        {
-          hostname: 'localhost',
-          port: 443,
-          path: '/api/v1/some/endpoint?limit=1',
-          agent: false,
-          timeout: 10,
-        },
-        (res) => {
-          if (res.statusCode !== 200) {
-            reject(new Error('Provider authentication failed'));
-          } else {
-            resolve();
-          }
-        },
+  private withBaseUri(path: string): string {
+    return `https://api.malwarebytes.com/${path}`;
+  }
+
+  private async request(
+    uri: string,
+    method: 'GET' | 'POST' | 'HEAD' = 'GET',
+    body?: any,
+  ): Promise<Response> {
+    if (!APIClient.accessToken) {
+      await this.getAuthenticationToken();
+    }
+
+    return await fetch(uri, {
+      method,
+      headers: {
+        Authorization: `Bearer ${APIClient.accessToken}`,
+        AccountId: this.accountId,
+        'Content-Type': 'application/json',
+      },
+      body: body || null,
+    });
+  }
+
+  private async makePaginatedRequest<K extends string, V>(
+    uri: string,
+    method: 'GET' | 'POST' | 'HEAD' = 'GET',
+    resource: K,
+    body?: any,
+  ): Promise<V[]> {
+    const results: V[] = [];
+    let nextCursor = '';
+
+    do {
+      const response = await this.request(
+        uri,
+        method,
+        JSON.stringify({
+          ...body,
+          page_size: 10,
```

Comment:
Oh, that's actually something I should have changed afterward. It was set to 10 initially just so that we have more pages in the responses - allowing us to work on/test pagination a little easier with fewer resources.

Do you have some preferences on what that number should be?

Reply:
The limit seems to be [0, 2000]. I've put 500 right now (We don't really know if you have some standard when it comes to this) but can be easily changed.

Reply:
Great - thank you for looking into the range. I think 500 is a totally reasonable response size. Thank you!